### ISSUE TITLE
Upgrade k8s failed

### ISSUE THREAD

**Issue by wuwo1952368901** [rancher/rke2#5062]
I modified the configuration of ETCD when creating the cluster, and there will be errors when upgrading to the k8s version. 
rke2-server.service log :

```
Unit rke2-server.service entered failed state.
Nov 21 11:29:37 master systemd[1]: rke2-server.service failed.
Nov 21 11:29:42 master systemd[1]: rke2-server.service holdoff time over, scheduling restart.
Nov 21 11:29:42 master systemd[1]: Stopped Rancher Kubernetes Engine v2 (server).
Nov 21 11:29:42 master systemd[1]: Starting Rancher Kubernetes Engine v2 (server)...
Nov 21 11:29:42 master sh[32693]: + /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
Nov 21 11:29:42 master sh[32693]: Failed to get unit file state for nm-cloud-setup.service: No such file or directory
Nov 21 11:29:42 master rke2[32700]: time="2023-11-21T11:29:42+08:00" level=warning msg="not running in CIS mode"
Nov 21 11:29:42 master rke2[32700]: time="2023-11-21T11:29:42+08:00" level=info msg="Starting rke2 v1.24.17+rke2r1 (2d67b55ca31e1ccb2c7c04320383ac6233acf3f5)"
Nov 21 11:29:42 master rke2[32700]: time="2023-11-21T11:29:42+08:00" level=info msg="Managed etcd cluster bootstrap already complete and initialized"
Nov 21 11:29:42 master rke2[32700]: time="2023-11-21T11:29:42+08:00" level=info msg="Starting temporary etcd to reconcile with datastore"
Nov 21 11:29:42 master rke2[32700]: {"level":"info","ts":"2023-11-21T11:29:42.392+0800","caller":"embed/etcd.go:131","msg":"configuring peer listeners","listen-peer-urls":["https://127.0.0.1:2380","https://172.20.0.161:2380","https://publicip:2380"]}
Nov 21 11:29:42 master rke2[32700]: {"level":"info","ts":"2023-11-21T11:29:42.392+0800","caller":"embed/etcd.go:368","msg":"closing etcd server","name":"master-7e846ac9","data-dir":"/var/lib/rancher/rke2/server/db/etcd-tmp","advertise-peer-urls":["http://127.0.0.1:2400"],"advertise-client-urls":["http://127.0.0.1:2399"]}
Nov 21 11:29:42 master rke2[32700]: {"level":"info","ts":"2023-11-21T11:29:42.392+0800","caller":"embed/etcd.go:370","msg":"closed etcd server","name":"master-7e846ac9","data-dir":"/var/lib/rancher/rke2/server/db/etcd-tmp","advertise-peer-urls":["http://127.0.0.1:2400"],"advertise-client-urls":["http://127.0.0.1:2399"]}
Nov 21 11:29:42 master rke2[32700]: time="2023-11-21T11:29:42+08:00" level=fatal msg="Failed to reconcile with temporary etcd: listen tcp 127.0.0.1:2380: bind: address already in use"
Nov 21 11:29:42 master systemd[1]: rke2-server.service: main process exited, code=exited, status=1/FAILURE
Nov 21 11:29:42 master systemd[1]: Failed to start Rancher Kubernetes Engine v2 (server).
```

My configuration is as follows:
```yaml
# /etc/rancher/rke2/config.yaml
etcd-arg:
  - "listen-client-urls=https://publicip:2379,https://172.20.0.161:2379,https://127.0.0.1:2379"
  - "listen-peer-urls=https://publicip:2380,https://172.20.0.161:2380,https://127.0.0.1:2380"
  - "quota-backend-bytes=5368709120"
```

**Comment by brandond** [comment]
Don't override the etcd options that configure the listen or advertise addresses, as this is very likely to break things or cause conflicts during startup. 

Also, 1.24 has been end of life for several months; I would recommend upgrading.

**Comment by wuwo1952368901** [comment]
Yes, I am trying to upgrade. Due to environmental issues, I have to modify the configuration of ETCD. Is there any other way to upgrade normally now?

**Comment by wuwo1952368901** [comment]
> Don't override the etcd options that configure the listen or advertise addresses, as this is very likely to break things or cause conflicts during startup.
> 
> Also, 1.24 has been end of life for several months; I would recommend upgrading.

Yes, I am trying to upgrade. Due to environmental issues, I have to modify the configuration of ETCD. Is there any other way to upgrade normally now?

**Comment by brandond** [comment]
We intentionally do not support configuring etcd to advertise or listen on anything other than the private IPs. RKE2 generates internal configuration for the addresses that ensures there are no port conflicts during startup; by overriding the configuration you are breaking that and as a result you will probably have to manually ensure that the etcd static pod is terminated (or the process stopped at the very least) when restarting the service.

**Comment by wuwo1952368901** [comment]
> We intentionally do not support configuring etcd to advertise or listen on anything other than the private IPs. RKE2 generates internal configuration for the addresses that ensures there are no port conflicts during startup; by overriding the configuration you are breaking that and as a result you will probably have to manually ensure that the etcd static pod is terminated (or the process stopped at the very least) when restarting the service.

After I stopped ETCD, I still couldn't upgrade normally.

Nov 21 15:37:22 master systemd[1]: Unit rke2-server.service entered failed state.
Nov 21 15:37:22 master systemd[1]: rke2-server.service failed.
Nov 21 15:37:27 master systemd[1]: rke2-server.service holdoff time over, scheduling restart.
Nov 21 15:37:27 master systemd[1]: Stopped Rancher Kubernetes Engine v2 (server).
Nov 21 15:37:27 master systemd[1]: Starting Rancher Kubernetes Engine v2 (server)...
Nov 21 15:37:27 master sh[18129]: + /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
Nov 21 15:37:27 master sh[18129]: Failed to get unit file state for nm-cloud-setup.service: No such file or directory
Nov 21 15:37:27 master rke2[18138]: time="2023-11-21T15:37:27+08:00" level=warning msg="not running in CIS mode"
Nov 21 15:37:27 master rke2[18138]: time="2023-11-21T15:37:27+08:00" level=info msg="Starting rke2 v1.24.17+rke2r1 (2d67b55ca31e1ccb2c7c04320383ac6233acf3f5)"
Nov 21 15:37:27 master rke2[18138]: time="2023-11-21T15:37:27+08:00" level=info msg="Managed etcd cluster bootstrap already complete and initialized"
Nov 21 15:37:27 master rke2[18138]: time="2023-11-21T15:37:27+08:00" level=info msg="Starting temporary etcd to reconcile with datastore"
Nov 21 15:37:27 master rke2[18138]: {"level":"info","ts":"2023-11-21T15:37:27.277+0800","caller":"embed/etcd.go:131","msg":"configuring peer listeners","listen-peer-urls":["https://0.0.0.0:2380"]}
Nov 21 15:37:27 master rke2[18138]: {"level":"info","ts":"2023-11-21T15:37:27.277+0800","caller":"embed/etcd.go:368","msg":"closing etcd server","name":"master-18c342da","data-dir":"/drive/lib/rancher/rke2/server/db/etcd-tmp","advertise-peer-urls":["http://127.0.0.1:2400"],"advertise-client-urls":["http://127.0.0.1:2399"]}
Nov 21 15:37:27 master rke2[18138]: {"level":"info","ts":"2023-11-21T15:37:27.277+0800","caller":"embed/etcd.go:370","msg":"closed etcd server","name":"master-18c342da","data-dir":"/drive/lib/rancher/rke2/server/db/etcd-tmp","advertise-peer-urls":["http://127.0.0.1:2400"],"advertise-client-urls":["http://127.0.0.1:2399"]}
Nov 21 15:37:27 master rke2[18138]: time="2023-11-21T15:37:27+08:00" level=fatal msg="Failed to reconcile with temporary etcd: cannot listen on TLS for [::]:2380: KeyFile and CertFile are not presented"
Nov 21 15:37:27 master systemd[1]: rke2-server.service: main process exited, code=exited, status=1/FAILURE

**Comment by brandond** [comment]
> My cluster nodes are on different networks and cannot be interconnected using the internal network. Therefore, I need to modify the ETCD configuration to create a cluster. In my case, how do I need to configure to achieve cluster upgrade?

This is intentionally not supported. I think you'll need to use `rke2 server --cluster-reset `on one of your nodes to reset the etcd cluster back to a single member, and run off a single server until you can configure your network such that your servers can reach each other by their private addresses.

> Does rke2 support configuring external etcd?

No, only internal etcd
